### DISCUSSION TITLE
Accessing and updating data | public / private collections

### DISCUSSION THREAD

**Discussion by VilleKokkarinen** [pocketbase/pocketbase#1978] [Show and tell]
I've got a forum / blog type of app under development, where i'd like to store images using pocketbase and fetch the url and only use the url in the app to manage the images.

Currently i don't think there is support for having a field hidden inside a Record, so one needs to split the collections into a private and a public collection.

ended up with the following solution:

- collection 'images'
fields : privatekey, data, url
![image](https://user-images.githubusercontent.com/20614421/222958656-9d26fbfb-606b-460b-9787-4334436ea61f.png)

-collection 'images_eol'
fields: eol, image relation(id & url)
![image](https://user-images.githubusercontent.com/20614421/222958796-18713895-6117-4642-a055-5610cb2808b9.png)

in the end i can now:
- give each authenticated user a private key, with which they can manage their uploaded images.
- every user can view the images via the image url.
- ability to manage the images' eol datetime, (maybe not everytime someone fetches one, but if someone likes a post, then loop all images in that post for example.)
- if a images' eol is reached, just delete / reduce it's size by 50%, and update eol for next time, so old images nobody really needs don't take up space.

**Though i was left wondering if it's possible to add a rate limiter, so i could disable image uploading if user has already uploaded let's say 100 images / x GB of data, for example?**

I think that solution is ok, and matches that of NoSQL patterns like firebase where one also has to divide documents into private /  public collections.

It would've been easier if i could just hide the privatekey from everyone by default from the images, and instead just have one collection, i think this was in the roadmap somewhere but couldn't really find from git.

Also if it was possible to keep same id in 2 different collections, i could've mapped both collections with the same record id, but i guess that breaks the unique id of pocketbase?

**Comment by ganigeorgiev** [comment]
> Currently i don't think there is support for having a field hidden inside a Record, so one needs to split the collections into a private and a public collection.

Yes, currently there is no built-in mechanism to hide a record field from the response. Depending on your use case, an alternative could be to create a read-only "View collection" (_[available with v0.13.0](https://github.com/pocketbase/pocketbase/releases/tag/v0.13.0)_) with only the subset of the fields.

> Though i was left wondering if it's possible to add a rate limiter, so i could disable image uploading if user has already uploaded let's say 100 images / x GB of data, for example?

This is too use-case specific. Generic rate limiter eventually will be added in the future as part of #1232, but for now remains a low priority. As a workaround you can use a reverse proxy (eg. nginx) and apply the rate limiter through it, although your use case is not a standard one (aka. `reqs/sec`), so I'm not sure if it can be achieved without some extra configuration.

---

> It would've been easier if i could just hide the privatekey from everyone by default from the images, and instead just have one collection, i think this was in the roadmap somewhere but couldn't really find from git.

I'm not sure that I understand your use case and why an extra "private key" is necessary for the users to manage their files.
If you are asking for private/protected files, this will be implemented soon (v0.15/v0.16) as part of #215 (_if you haven't already, you can check the current tasks priority in our [roadmap](https://github.com/orgs/pocketbase/projects/2)_).

---

> Also if it was possible to keep same id in 2 different collections, i could've mapped both collections with the same record id, but i guess that breaks the unique id of pocketbase?

You can use the same ids for records from different collections (_each collection is represented as a separate SQLite table_).

**Reply by VilleKokkarinen** [reply]
i wanted to introduce a private key, since in my case the user's id is in their public collection.
any bad actor could pretty much just list all user ids, and then delete all images using those
(if the key field was 'author_id' or something)

Oh, i did get the same id mapping working, i guess i typo'd something while testing 👍, saves extending the relation if i don't need to!

But yeah, Pocketbase is looking like a very good backend solution atleast for the scale i had in mind.
Looking forward for the next releases 👍

**Reply by ganigeorgiev** [reply]
> i wanted to introduce a private key, since in my case the user's id is in their public collection.
any bad actor could pretty much just list all user ids, and then delete all images using those
(if the key field was 'author_id' or something)

I'm still not sure that I understand what do you mean. You can restrict access to only the authenticated owner of a record using the `author_id = @request.auth.id` rule (_assuming that `author_id` is  nonempty single `relation` field to the `users` collection_).

**Reply by VilleKokkarinen** [reply]
Currently i'm not using pocketbase auth, just managing images / files in the db.
No auth / user data (for now).

Might migrate authentication over in the future.
After that of course it would work by using pocketbase authentication.
